### ISSUE TITLE
Problem with BOSH-Links and errands: Can't generate network settings without an IP

### ISSUE THREAD

**Issue by JamesClonk** [cloudfoundry/bosh#1977]
BOSH version: `266.2.0 and 266.3.0`
Deployment: `cf-mysql`

Since upgrading the BOSH director from the 265.2.0 to the latest 266.x.0 series there has been a problem with BOSH-links in combination with errands it seems.

The cf-mysql-release has a `mysql` job with the following spec:
```
...
provides:
- name: mysql
  type: mysql
  properties:
  ...
```
It also has an errand `bootstrap` which consumes the `mysql` link:
```
consumes:
- name: mysql
  type: mysql
  optional: true
```

The deployment manifest we use specifies the `provides` part for the mysql job, and the  `consumes` part for the bootstrap errand, to have it consume the mysql instance-group.

All has worked fine with BOSH 265.2.0, but since the 266.x.0 series of the BOSH director deployments fail with the following error:
```
Task 112201 | 08:38:12 | Preparing deployment: Preparing deployment (00:00:11)
                      L Error: Can't generate network settings without an IP
Task 112201 | 08:38:23 | Error: Can't generate network settings without an IP
```
The deployment manifest hasn't changed though, release version of cf-mysql-release stayed the same too.
Something about BOSH-links in combination with errands must have gone broken. The deployment goes through fine as soon as I remove errand from it (or remove the BOSH-link `consumes` from the errand job definition).
If did some testing and pinned it down exactly to the consumption of BOSH-links in errand jobs. For some reason the new version of BOSH beginning with 266.2.0 fails deployments with those!

Full debug log:
```
D, [2018-06-14T08:38:23.959120 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000068s) (conn: 69906706277620) BEGIN
D, [2018-06-14T08:38:23.959830 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000250s) (conn: 69906706277620) UPDATE `link_provider_intents` SET `link_provider_id` = 145, `original_name` = 'arbitrator', `type` = 'arbitrator', `name` = 'arbitrator', `content` = '{\"deployment_name\":\"galera\",\"domain\":\"bosh\",\"default_network\":\"management\",\"networks\":[\"management\"],\"instance_group\":\"arbitrator\",\"properties\":{},\"use_short_dns_addresses\":false,\"use_dns_addresses\":false,\"instances\":[]}', `shared` = 0, `consumable` = 1, `metadata` = '{\"mapped_properties\":{}}', `serial_id` = 2 WHERE (`id` = 256) LIMIT 1
D, [2018-06-14T08:38:23.960689 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000718s) (conn: 69906706277620) COMMIT
D, [2018-06-14T08:38:23.961309 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000081s) (conn: 47100864174660) SELECT * FROM `local_dns_encoded_azs`
D, [2018-06-14T08:38:23.962152 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000111s) (conn: 69906704468040) SELECT `local_dns_encoded_instance_groups`.`id` AS `id`, `local_dns_encoded_instance_groups`.`name` AS `name`, `deployments`.`name` AS `deployment_name` FROM `local_dns_encoded_instance_groups` INNER JOIN `deployments` ON (`local_dns_encoded_instance_groups`.`deployment_id` = `deployments`.`id`)
D, [2018-06-14T08:38:23.964139 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000080s) (conn: 69906706297900) BEGIN
D, [2018-06-14T08:38:23.964762 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000300s) (conn: 69906706297900) UPDATE `tasks` SET `event_output` = CONCAT(`event_output`, '{\"time\":1528965503,\"stage\":\"Preparing deployment\",\"tags\":[],\"total\":1,\"task\":\"Preparing deployment\",\"index\":1,\"state\":\"failed\",\"progress\":100,\"data\":{\"error\":\"Can\'t generate network settings without an IP\"}}\n') WHERE (`id` = 112201)
D, [2018-06-14T08:38:23.965390 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000532s) (conn: 69906706297900) COMMIT
D, [2018-06-14T08:38:23.966958 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000637s) (conn: 69906704468040) DELETE FROM `locks` WHERE ((`name` = 'lock:deployment:galera') AND (`uid` = '44d4edcd-d5bb-4f34-bf96-5925be805edc'))
D, [2018-06-14T08:38:23.967036 #32627] [task:112201] DEBUG -- DirectorJobRunner: Deleted lock: lock:deployment:galera uid: 44d4edcd-d5bb-4f34-bf96-5925be805edc
D, [2018-06-14T08:38:23.967815 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000068s) (conn: 69906706297900) BEGIN
D, [2018-06-14T08:38:23.968389 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000149s) (conn: 69906706297900) INSERT INTO `events` (`parent_id`, `timestamp`, `user`, `action`, `object_type`, `object_name`, `error`, `task`, `deployment`, `instance`, `context_json`) VALUES (NULL, '2018-06-14 08:38:23', 'ci', 'release', 'lock', 'lock:deployment:galera', NULL, '112201', 'galera', NULL, '{}')
D, [2018-06-14T08:38:23.968775 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000169s) (conn: 69906706297900) SELECT * FROM `events` WHERE `id` = 68597273
D, [2018-06-14T08:38:23.969411 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000450s) (conn: 69906706297900) COMMIT
I, [2018-06-14T08:38:23.969800 #32627] [task:112201]  INFO -- DirectorJobRunner: sending update deployment error event
D, [2018-06-14T08:38:23.969899 #32627] [task:112201] DEBUG -- DirectorJobRunner: SENT: hm.director.alert {"id":"b6931290-0a18-415a-b53d-e4c9e5d3d169","severity":3,"source":"director","title":"director - error during update deployment","summary":"Error during update deployment for 'galera' against Director '6eeddaef-9dd9-4614-ba6e-ac33a7f21a03': #<Bosh::Director::NetworkReservationIpMissing: Can't generate network settings without an IP>","created_at":1528965503,"deployment":"galera"}
D, [2018-06-14T08:38:23.970684 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000068s) (conn: 69906706277620) BEGIN
D, [2018-06-14T08:38:23.971169 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000157s) (conn: 69906706277620) INSERT INTO `events` (`parent_id`, `timestamp`, `user`, `action`, `object_type`, `object_name`, `error`, `task`, `deployment`, `instance`, `context_json`) VALUES (68597249, '2018-06-14 08:38:23', 'ci', 'update', 'deployment', 'galera', 'Can\'t generate network settings without an IP', '112201', 'galera', NULL, '{}')
D, [2018-06-14T08:38:23.971561 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000212s) (conn: 69906706277620) SELECT * FROM `events` WHERE `id` = 68597274
D, [2018-06-14T08:38:23.973483 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.001730s) (conn: 69906706277620) COMMIT
D, [2018-06-14T08:38:23.974607 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000156s) (conn: 69906704468040) SELECT * FROM `deployments` WHERE (`name` = 'galera') LIMIT 1
D, [2018-06-14T08:38:23.976992 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000157s) (conn: 69906706297900) SELECT * FROM `variable_sets` WHERE (`variable_sets`.`deployment_id` = 77) ORDER BY `created_at` DESC LIMIT 1
D, [2018-06-14T08:38:23.977644 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000075s) (conn: 69906706277620) BEGIN
D, [2018-06-14T08:38:23.978285 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000195s) (conn: 69906706277620) UPDATE `variable_sets` SET `writable` = 0 WHERE (`id` = 1042) LIMIT 1
D, [2018-06-14T08:38:23.979189 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000757s) (conn: 69906706277620) COMMIT
D, [2018-06-14T08:38:23.979728 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000068s) (conn: 47100864174660) BEGIN
D, [2018-06-14T08:38:23.980232 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000201s) (conn: 47100864174660) UPDATE `tasks` SET `event_output` = CONCAT(`event_output`, '{\"time\":1528965503,\"error\":{\"code\":130005,\"message\":\"Can\'t generate network settings without an IP\"}}\n') WHERE (`id` = 112201)
D, [2018-06-14T08:38:23.980830 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000495s) (conn: 47100864174660) COMMIT
E, [2018-06-14T08:38:23.980998 #32627] [task:112201] ERROR -- DirectorJobRunner: Can't generate network settings without an IP
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/manual_network.rb:43:in `network_settings'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/network_settings.rb:25:in `block in to_hash'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/network_settings.rb:23:in `each'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/network_settings.rb:23:in `to_hash'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/network_settings.rb:52:in `network_address'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/instance_plan.rb:268:in `network_address'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/links/link.rb:33:in `block in spec'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/links/link.rb:24:in `map'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/links/link.rb:24:in `spec'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/links/links_manager.rb:288:in `block (2 levels) in update_provider_intents_contents'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/links/links_manager.rb:281:in `each'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/links/links_manager.rb:281:in `block in update_provider_intents_contents'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/links/links_manager.rb:279:in `each'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/links/links_manager.rb:279:in `update_provider_intents_contents'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/assembler.rb:151:in `bind_links'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/assembler.rb:88:in `bind_models'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/update_deployment.rb:91:in `block (2 levels) in perform'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/event_log.rb:105:in `advance_and_track'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/update_deployment.rb:79:in `block in perform'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/lock_helper.rb:13:in `block in with_deployment_lock'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/lock.rb:79:in `lock'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/lock_helper.rb:13:in `with_deployment_lock'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/update_deployment.rb:58:in `perform'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/job_runner.rb:99:in `perform_job'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/job_runner.rb:34:in `block in run'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh_common-0.0.0/lib/common/thread_formatter.rb:52:in `with_thread_name'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/job_runner.rb:34:in `run'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/base_job.rb:10:in `perform'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/db_job.rb:36:in `block in perform'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/bosh-director-0.0.0/lib/bosh/director/jobs/db_job.rb:83:in `block (3 levels) in run'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/eventmachine-1.2.5/lib/eventmachine.rb:1076:in `block in spawn_threadpool'
/var/vcap/data/packages/director/4d1cc6d9fe5c6bb2cdd1c5a6d225ecdf3a674690/gem_home/ruby/2.4.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
D, [2018-06-14T08:38:23.981589 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000174s) (conn: 69906704468040) SELECT * FROM `tasks` WHERE `id` = 112201
D, [2018-06-14T08:38:23.982240 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000063s) (conn: 69906706297900) BEGIN
D, [2018-06-14T08:38:23.983035 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000279s) (conn: 69906706297900) UPDATE `tasks` SET `state` = 'error', `timestamp` = '2018-06-14 08:38:23', `description` = 'create deployment', `result` = 'Can\'t generate network settings without an IP', `output` = '/var/vcap/store/director/tasks/112201', `checkpoint_time` = '2018-06-14 08:38:11', `type` = 'update_deployment', `username` = 'ci', `deployment_name` = 'galera', `started_at` = '2018-06-14 08:38:11', `event_output` = '{\"time\":1528965492,\"stage\":\"Preparing deployment\",\"tags\":[],\"total\":1,\"task\":\"Preparing deployment\",\"index\":1,\"state\":\"started\",\"progress\":0}\n{\"time\":1528965503,\"stage\":\"Preparing deployment\",\"tags\":[],\"total\":1,\"task\":\"Preparing deployment\",\"index\":1,\"state\":\"failed\",\"progress\":100,\"data\":{\"error\":\"Can\'t generate network settings without an IP\"}}\n{\"time\":1528965503,\"error\":{\"code\":130005,\"message\":\"Can\'t generate network settings without an IP\"}}\n', `result_output` = '', `context_id` = '' WHERE (`id` = 112201) LIMIT 1
D, [2018-06-14T08:38:23.983665 #32627] [task:112201] DEBUG -- DirectorJobRunner: (0.000469s) (conn: 69906706297900) COMMIT
I, [2018-06-14T08:38:23.983809 #32627] []  INFO -- DirectorJobRunner: Task took 12.267762669 seconds to process.
```

Any deployment having an errand consuming BOSH-links is failing.

**Comment by xtreme-andrew-su** [comment]
Hey @JamesClonk, I want to confirm, are there any other changes to your manifest between your usage on BOSH 265 and 266?

Can you provide me your manifest before and after your link changes? Are there any changes that was done to your cloud-config? More specifically your networking.

**Comment by JamesClonk** [comment]
Hi @xtreme-andrew-su,
No, there are no changes in the deployment manifest between 265 and 266. In fact I can download an already deployed manifest (that was previously deployed with 265 or earlier) and try to redeploy it with 266 and it will fail with above mentioned error message.
Just verified that once more with another deployment that contains an errand and hasn't been touched in a while:
```
$ bosh -d docker-registry manifest > test.yml
$ bosh -n -d docker-registry deploy test.yml --no-redact
...
Task 112701 | 17:33:58 | Preparing deployment: Preparing deployment (00:00:12)
                      L Error: Can't generate network settings without an IP
Task 112701 | 17:34:10 | Error: Can't generate network settings without an IP
```
😩 

There's not much to see in the manifest. There's simply an errand that consumes a BOSH-link from another instance-group in the same deployment. The deployment fails until I remove the errand itself or remove the `consumes` from it (but then the errand will fail because it needs to know about the other instance-group).
No, no changes done to cloud-config between 265 and 266, at least regarding networking. There might have been some small changes to some vm-types, but unrelated to and not affecting that particular deployments vm-types.


Anyway, here's a distilled down example I was using for testing the problem:
```
instance_groups:
- name: server
  instances: 3
  networks:
  - default:
    - dns
    - gateway
    name: test
  stemcell: default
  vm_type: small
  jobs:
  - name: consul
    release: consul
    consumes:
      consul:
        deployment: consul-test
        from: consul
    provides:
      consul:
        as: consul
    properties: { ... lots of stuff here ... }

- name: agent
  instances: 1
  lifecycle: errand
  networks:
  - default:
    - dns
    - gateway
    name: test
  stemcell: default
  vm_type: errand
  jobs:
  - name: smoke-tests
    release: consul
    consumes:
      consul:
        deployment: consul-test
        from: consul
    properties: {}
```

(note: that's our own internal custom Consul release, not the one used by Cloud Foundry. It has a job `consul` which provides and consumes itself to know about its peers, and a smoke-test errand that also wants to consume the consul job to find the server-nodes)
As soon as I remove the errand from the deployment manifest it will work on BOSH 266. Until then it fails. Works fine either way on BOSH 265 or earlier.

**Comment by xtreme-andrew-su** [comment]
From your example, I created a sample [test-release](https://github.com/xtreme-andrew-su/test-release). 

Creating a simple bosh-lite warden director using the latest [bosh-deployment v266.2.0](https://github.com/cloudfoundry/bosh-deployment) then deployed using this [manifest](https://github.com/xtreme-andrew-su/test-release/blob/master/manifests/manifest.yml) which as far as I can tell is the same as your example.

I was not able to reproduce this issue.

I've included a few other manifest which shows the issue you're seeing but only for providers that are errands.

You can reach out to me on https://cloudfoundry.slack.com/messages/bosh for `@asu`.

**Comment by JamesClonk** [comment]
@xtreme-andrew-su 

**TLDR**: https://github.com/xtreme-andrew-su/test-release/pull/1

---

I was looking at your test-release, and the first I thought about was adding an erb template to the consumer job that will actually use the link to get the providers IP address, thinking that that would cause it to misbehave:
```
templates:
  test.yml.erb: config/test.yml
```
```
<%=
  provider_link = link('provider')
  provider_ips = provider_link.instances.map { |instance| instance.address }
  config = { 
    IPs: provider_ips.map {|ip| "http://#{ip}" },
  }
  JSON.pretty_generate(config)
%>
```
But to my surprise it still worked. The BOSH director 266.x.0 that I'm using that is giving me trouble with these other errands was able to deploy your test-release just fine.


So I've tried again with one of my own deployments, but this time I was thinking along the lines that maybe the reason for it working is that it was a completely new deployment, not updating an existing one.
I downloaded an existing manifest, modified the deployment name, and tried deploying it:
```
$ bosh -d docker-registry manifest > docker-2.yml
$ vi docker-2.yml # modifying name to docker-2, nothing else..
$ bosh -d docker-2 deploy docker-2.yml
...
Task 112790 | 08:24:14 | Preparing deployment: Preparing deployment (00:00:11)
                      L Error: Can't generate network settings without an IP
Task 112790 | 08:24:25 | Error: Can't generate network settings without an IP
```
Same error message.

This is weird. So I added the same consul job that's inside my docker-registry's errand to your test-release deployment, to consume the consul link. Then deployed it again:
```
Using deployment 'test_dep'

  releases:
+ - name: consul
+   version: 2.0.9
  
  instance_groups:
  - name: provider_ig
    jobs:
    - name: provider
+     release: test
  - name: consumer_ig
    jobs:
    - name: consumer
+     release: test
+   - consumes:
+       consul:
+         deployment: consul
+         from: consul
+     name: consul
+     properties: {}
+     release: consul
Task 112793

Task 112793 | 08:28:34 | Preparing deployment: Preparing deployment (00:00:11)
                      L Error: Can't generate network settings without an IP
Task 112793 | 08:28:45 | Error: Can't generate network settings without an IP
```

Now I'm thinking that it might be that the main consul deployment providing the link is somehow broken. So I redeploy that one hoping it would fix the issue:
```
$ bosh -d consul manifest > consul.yml
$ bosh -d consul deploy consul.yml
...
Task 112794 done
```
Trying again the `test_dep`:
```
$ bosh -n -d test_dep deploy manifests/manifest.yml 
Using deployment 'test_dep'

  releases:
+ - name: consul
+   version: 2.0.9
  
  instance_groups:
  - name: provider_ig
    jobs:
    - name: provider
+     release: test
  - name: consumer_ig
    jobs:
    - name: consumer
+     release: test
+   - consumes:
+       consul:
+         deployment: consul
+         from: consul
+     name: consul
+     properties: {}
+     release: consul
Task 112796

Task 112796 | 08:45:05 | Preparing deployment: Preparing deployment (00:00:12)
                      L Error: Can't generate network settings without an IP
Task 112796 | 08:45:17 | Error: Can't generate network settings without an IP

Task 112796 Started  Sat Jun 16 08:45:05 UTC 2018
Task 112796 Finished Sat Jun 16 08:45:17 UTC 2018
Task 112796 Duration 00:00:12
Task 112796 error

Updating deployment:
  Expected task '112796' to succeed but state is 'error'

Exit code 1
```
Still no luck! :(

Tried some more things, like remote the `lifecycle: errand` from the `consumer` job, then the deployment will work.
Or moving the consul job from the consumer errand to the provider job, then the deployment will also work.

So it boils down to the `consul` job from our `consul` release containing something that makes it fail when used in an errand, starting with BOSH directo 266.x.0 series (It still works completely fine on BOSH 265.x or earlier).

After some more digging and comparing the differences between your `consumer` job and our `consul` job I finally found the cause.
The job has a `consumes` and `provides` of the same link at the same time, with the consumes being marked as `optional`.
This used to perfectly work on earlier BOSH director versions.
I've made a PR for your test-release (https://github.com/xtreme-andrew-su/test-release/pull/1) where you can check it out. It contains the new job `both` and an example manifest `manifest-both.yml`.

Deploying the manifest-both.yml works on BOSH 265 or earlier, tested it with this older BOSH here:
```
Name      test-bosh  
UUID      f3abb886-deee-426f-956b-71431826d8ac  
Version   264.6.0 (00000000)  
CPI       vsphere_cpi  
Features  compiled_package_cache: enabled  
          config_server: disabled  
          dns: disabled  
          snapshots: disabled
```
Worked without a hitch.
```
Task 726 | 09:21:16 | Preparing deployment: Preparing deployment (00:00:02)
Task 726 | 09:21:18 | Preparing package compilation: Finding packages to compile (00:00:00)
Task 726 | 09:21:18 | Creating missing vms: provider_ig/cdf8cc86-7190-44f9-9f81-039640c274ae (0) (00:01:03)
Task 726 | 09:22:22 | Updating instance provider_ig: provider_ig/cdf8cc86-7190-44f9-9f81-039640c274ae (0) (canary) (00:00:28)

Task 726 Started  Sat Jun 16 09:21:16 UTC 2018
Task 726 Finished Sat Jun 16 09:22:50 UTC 2018
Task 726 Duration 00:01:34
Task 726 done

Succeeded
```

On a 266.x BOSH it fails with `Error: Can't generate network settings without an IP`:
```
Task 112816 | 09:26:34 | Preparing deployment: Preparing deployment (00:00:11)
                      L Error: Can't generate network settings without an IP
Task 112816 | 09:26:45 | Error: Can't generate network settings without an IP

Task 112816 Started  Sat Jun 16 09:26:34 UTC 2018
Task 112816 Finished Sat Jun 16 09:26:45 UTC 2018
Task 112816 Duration 00:00:11
Task 112816 error

Updating deployment:
  Expected task '112816' to succeed but state is 'error'

Exit code 1
```

Same release, same manifest..

**Comment by xtreme-andrew-su** [comment]
@JamesClonk FYI
Related stories:
- https://www.pivotaltracker.com/story/show/158444025
- https://www.pivotaltracker.com/story/show/158444050

**Comment by JamesClonk** [comment]
@xtreme-andrew-su cool, thanks! 😃

**Comment by freddesbiens** [comment]
@JamesClonk We implemented the first of the two stories mentioned by @xtreme-andrew-su above. The second one is much more complicated, however, and it will probably stay buried in our backlog forever since stories with more value with always take precedence. A possible workaround is to leverage co-located errands; the errand is guaranteed to have an IP address then, since it is paired with a running instance.

Closing this issue for now. Thank you for your patience.
